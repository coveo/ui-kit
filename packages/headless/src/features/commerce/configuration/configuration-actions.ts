import {StringValue} from '@coveo/bueno';
import {createAction} from '@reduxjs/toolkit';
import {PlatformEnvironment} from '../../../utils/url-utils.js';
import {
  nonEmptyString,
  validatePayload,
} from '../../../utils/validate-payload.js';
import {
  UpdateAnalyticsConfigurationActionCreatorPayload,
  UpdateBasicConfigurationActionCreatorPayload,
  analyticsConfigurationSchema,
} from '../../configuration/configuration-actions.js';

export type UpdateBasicConfigurationPayload =
  UpdateBasicConfigurationActionCreatorPayload;

export const updateBasicConfiguration = createAction(
  'commerce/configuration/updateBasicConfiguration',
  (payload: UpdateBasicConfigurationPayload) =>
    validatePayload(payload, {
      accessToken: nonEmptyString,
      environment: new StringValue<PlatformEnvironment>({
        required: false,
        constrainTo: ['prod', 'hipaa', 'stg', 'dev'],
      }),
      organizationId: nonEmptyString,
    })
);

export type UpdateProxyBaseUrlPayload = {
  /**
   * The base URL to use to proxy Coveo commerce requests (e.g., `https://example.com/commerce`).
   *
   * This is an advanced option that you should only set if you need to proxy Coveo commerce requests through your own
   * server. In most cases, you should not set this option.
   *
   * By default, no proxy is used and the Coveo commerce requests are sent directly to the Coveo platform through the
   * platform [organization endpoint](https://docs.coveo.com/en/mcc80216) resolved from the `organizationId` and
   * `environment` values provided in your engine configuration (i.e.,
   * `https://<organizationId>.org.coveo.com` or
   * `https://<organizationId>.org<environment>.coveo.com`, if the `environment` values is specified and
   * different from `prod`).
   *
   * If you set this option, you must also implement the following proxy endpoints on your server, otherwise the
   * commerce engine will not work properly:
   *
   * - `POST` `/facet` to proxy requests to [`POST` `https://<organizationId>.org<environment|>.coveo.com/rest/organizations/<organizationId>/commerce/v2/facet`](https://docs.coveo.com/en/103/api-reference/commerce-api#tag/Facet/operation/facet)
   * - `POST` `/listing` to proxy requests to [`POST` `https://<organizationId>.org<environment|>.coveo.com/rest/organizations/<organizationId>/commerce/v2/listing`](https://docs.coveo.com/en/103/api-reference/commerce-api#tag/Listings/operation/getListing)
   * - `POST` `/productSuggest` to proxy requests to [`POST` `https://<organizationId>.org<environment|>.coveo.com/rest/organizations/<organizationId>/commerce/v2/productSuggest`](https://docs.coveo.com/en/103/api-reference/commerce-api#tag/Search/operation/productSuggest)
   * - `POST` `/querySuggest` to proxy requests to [`POST` `https://<organizationId>.org<environment|>.coveo.com/rest/organizations/<organizationId>/commerce/v2/querySuggest`](https://docs.coveo.com/en/103/api-reference/commerce-api#tag/Search/operation/querySuggest)
   * - `POST` `/recommendations` to proxy requests to [`POST` `https://<organizationId>.org<environment|>.coveo.com/rest/organizations/<organizationId>/commerce/v2/recommendations`](https://docs.coveo.com/en/103/api-reference/commerce-api#tag/Recommendations/operation/recommendations)
   * - `POST` `/search` to proxy requests to [`POST` `https://<organizationId>.org<environment|>.coveo.com/rest/organizations/<organizationId>/commerce/v2/search`](https://docs.coveo.com/en/103/api-reference/commerce-api#tag/Search/operation/search)
   **/
  proxyBaseUrl?: string;
};

export const updateProxyBaseUrl = createAction(
  'commerce/configuration/updateProxyBaseUrl',
  (payload: UpdateProxyBaseUrlPayload) =>
    validatePayload(payload, {
      proxyBaseUrl: new StringValue({required: false, url: true}),
    })
);

export type UpdateAnalyticsConfigurationPayload = Pick<
  UpdateAnalyticsConfigurationActionCreatorPayload,
  'enabled' | 'proxyBaseUrl' | 'source' | 'trackingId'
>;

export const updateAnalyticsConfiguration = createAction(
  'commerce/configuration/updateAnalyticsConfiguration',
  (payload: UpdateAnalyticsConfigurationPayload) => {
    return validatePayload(payload, {
      enabled: analyticsConfigurationSchema.enabled,
      proxyBaseUrl: analyticsConfigurationSchema.proxyBaseUrl,
      source: analyticsConfigurationSchema.source,
      trackingId: analyticsConfigurationSchema.trackingId,
    });
  }
);

export const disableAnalytics = createAction(
  'commerce/configuration/analytics/disable'
);

export const enableAnalytics = createAction(
  'commerce/configuration/analytics/enable'
);
