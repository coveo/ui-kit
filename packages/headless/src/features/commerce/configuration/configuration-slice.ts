import {isNullOrUndefined} from '@coveo/bueno';
import {createReducer} from '@reduxjs/toolkit';
import {
  disableAnalytics,
  enableAnalytics,
  updateAnalyticsConfiguration,
  UpdateAnalyticsConfigurationPayload,
  updateBasicConfiguration,
  UpdateBasicConfigurationPayload,
  updateProxyBaseUrl,
  UpdateProxyBaseUrlPayload,
} from './configuration-actions';
import {
  ConfigurationState,
  getConfigurationInitialState,
} from './configuration-state';

export const configurationReducer = createReducer(
  getConfigurationInitialState(),
  (builder) =>
    builder
      .addCase(updateBasicConfiguration, (state, action) => {
        handleUpdateBasicConfiguration(state, action.payload);
      })
      .addCase(updateProxyBaseUrl, (state, action) => {
        handleUpdateCommerceProxyBaseUrl(state, action.payload);
      })
      .addCase(updateAnalyticsConfiguration, (state, action) => {
        handleUpdateCommerceAnalyticsConfiguration(state, action.payload);
      })
      .addCase(disableAnalytics, (state) => {
        state.analytics.enabled = false;
      })
      .addCase(enableAnalytics, (state) => {
        state.analytics.enabled = true;
      })
);

function handleUpdateBasicConfiguration(
  state: ConfigurationState,
  payload: UpdateBasicConfigurationPayload
) {
  if (!isNullOrUndefined(payload.accessToken)) {
    state.accessToken = payload.accessToken;
  }

  state.environment = payload.environment ?? 'prod';

  if (!isNullOrUndefined(payload.organizationId)) {
    state.organizationId = payload.organizationId;
  }
}

function handleUpdateCommerceProxyBaseUrl(
  state: ConfigurationState,
  payload: UpdateProxyBaseUrlPayload
) {
  if (!isNullOrUndefined(payload.proxyBaseUrl)) {
    state.commerce.apiBaseUrl = payload.proxyBaseUrl;
  }
}

function handleUpdateCommerceAnalyticsConfiguration(
  state: ConfigurationState,
  payload: UpdateAnalyticsConfigurationPayload
) {
  if (!isNullOrUndefined(payload.enabled)) {
    state.analytics.enabled = payload.enabled;
  }
  if (!isNullOrUndefined(payload.proxyBaseUrl)) {
    state.analytics.apiBaseUrl = payload.proxyBaseUrl;
  }
  if (!isNullOrUndefined(payload.source)) {
    state.analytics.source = payload.source;
  }
  if (!isNullOrUndefined(payload.trackingId)) {
    state.analytics.trackingId = payload.trackingId;
  }
}
