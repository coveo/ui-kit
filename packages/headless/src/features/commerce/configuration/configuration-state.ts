import {CoreConfigurationState} from '../../configuration/configuration-state.js';

export interface ConfigurationState extends CoreConfigurationState {
  /**
   * The global headless engine Commerce API configuration.
   */
  commerce: CommerceState;
}

interface CommerceState {
  /**
   * The Commerce API base URL to use.
   *
   * By default, will append `/rest/organizations/{organizationId}/commerce/v2` to the automatically resolved
   * platform [organization endpoint](https://docs.coveo.com/en/mcc80216)
   * (i.e., `https;://<ORG_ID>.org<hipaa|dev|stg|>.coveo.com`)
   *
   * If necessary, you can override this value by specifying a `commerce.proxyBaseUrl` in the configuration of your
   * commerce engine, or by manually dispatching the `updateProxyBaseUrl` action.
   */
  apiBaseUrl?: string;
}

export const getConfigurationInitialState: () => ConfigurationState = () => ({
  accessToken: '',
  environment: 'prod',
  organizationId: '',
  analytics: {
    enabled: true,
    trackingId: '',
    source: {},
  },
  commerce: {},
});
