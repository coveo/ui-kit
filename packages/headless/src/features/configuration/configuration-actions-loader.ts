import {PayloadAction} from '@reduxjs/toolkit';
import {CoreEngine} from '../../app/engine';
import {configuration} from '../../app/reducers';
import {
  disableAnalytics,
  enableAnalytics,
  setOriginLevel2,
  SetOriginLevel2ActionCreatorPayload,
  setOriginLevel3,
  SetOriginLevel3ActionCreatorPayload,
  updateAnalyticsConfiguration,
  UpdateAnalyticsConfigurationActionCreatorPayload,
  updateBasicConfiguration,
  UpdateBasicConfigurationActionCreatorPayload,
  AnalyticsRuntimeEnvironment,
} from './configuration-actions';

export {
  SetOriginLevel2ActionCreatorPayload,
  SetOriginLevel3ActionCreatorPayload,
  UpdateAnalyticsConfigurationActionCreatorPayload,
  AnalyticsRuntimeEnvironment,
  UpdateBasicConfigurationActionCreatorPayload,
};

/**
 * The configuration action creators.
 */
export interface ConfigurationActionCreators {
  /**
   * Disables analytics tracking.
   *
   * @returns A dispatchable action.
   */
  disableAnalytics(): PayloadAction;

  /**
   * Enables analytics tracking.
   *
   * @returns A dispatchable action.
   */
  enableAnalytics(): PayloadAction;

  /**
   * Sets originLevel2 for analytics tracking.
   *
   * @param payload - The action creator payload.
   * @returns A dispatchable action.
   */
  setOriginLevel2(
    payload: SetOriginLevel2ActionCreatorPayload
  ): PayloadAction<SetOriginLevel2ActionCreatorPayload>;

  /**
   * Sets originLevel3 for analytics tracking.
   *
   * @param payload - The action creator payload.
   * @returns A dispatchable action.
   * */
  setOriginLevel3(
    payload: SetOriginLevel3ActionCreatorPayload
  ): PayloadAction<SetOriginLevel3ActionCreatorPayload>;

  /**
   * Updates the analytics configuration.
   *
   * @param payload - The action creator payload.
   * @returns A dispatchable action.
   * */
  updateAnalyticsConfiguration(
    payload: UpdateAnalyticsConfigurationActionCreatorPayload
  ): PayloadAction<UpdateAnalyticsConfigurationActionCreatorPayload>;

  /**
   * Updates the global headless engine configuration.
   *
   * @param payload - The action creator payload.
   * @returns A dispatchable action.
   * */
  updateBasicConfiguration(
    payload: UpdateBasicConfigurationActionCreatorPayload
  ): PayloadAction<UpdateBasicConfigurationActionCreatorPayload>;
}

/**
 * Loads the `configuration` reducer and returns possible action creators.
 *
 * @param engine - The headless engine.
 * @returns An object holding the action creators.
 */
export function loadConfigurationActions(
  engine: CoreEngine
): ConfigurationActionCreators {
  engine.addReducers({configuration});

  return {
    disableAnalytics,
    enableAnalytics,
    setOriginLevel2,
    setOriginLevel3,
    updateAnalyticsConfiguration,
    updateBasicConfiguration,
  };
}
