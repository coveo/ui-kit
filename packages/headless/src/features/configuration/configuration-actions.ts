import {createAction, createAsyncThunk} from '@reduxjs/toolkit';
import {validatePayload} from '../../utils/validate-payload';
import {StringValue, BooleanValue, Value} from '@coveo/bueno';
import {IRuntimeEnvironment} from 'coveo.analytics';

export const localeValidation = new StringValue({
  emptyAllowed: false,
  required: false,
  regex: /[a-z]{2}-[A-Z]{2}/,
});

const originSchemaOnConfigUpdate = () =>
  new StringValue({emptyAllowed: false, required: false});

const originSchemaOnUpdate = () =>
  new StringValue({emptyAllowed: false, required: true});

/**
 * Updates the global headless engine configuration.
 * @param accessToken (string) The access token to use to authenticate requests against the Coveo Cloud endpoints. Typically, this will be an API key or search token that grants the privileges to execute queries and push usage analytics data in the target Coveo Cloud organization.
 * @param organizationId (string) The unique identifier of the target Coveo Cloud organization (e.g., `mycoveocloudorganizationg8tp8wu3`)
 * @param platformUrl (string) The Plaform URL to use (e.g., `https://platform.cloud.coveo.com`).
 */
export const updateBasicConfiguration = createAction(
  'configuration/updateBasicConfiguration',
  (payload: {
    accessToken?: string;
    organizationId?: string;
    platformUrl?: string;
  }) =>
    validatePayload(payload, {
      accessToken: new StringValue({emptyAllowed: false}),
      organizationId: new StringValue({emptyAllowed: false}),
      platformUrl: new StringValue({url: true, emptyAllowed: false}),
    })
);

/**
 * Updates the search configuration.
 * @param apiBaseUrl (string) The Search API base URL to use (e.g., `https://platform.cloud.coveo.com/rest/search/v2`).
 * @param pipeline (string) The name of the query pipeline to use for the query (e.g., `External Search`). If not specified, the default query pipeline will be used.
 * @param searchHub (string) The first level of origin of the request, typically the identifier of the graphical search interface from which the request originates (e.g., `ExternalSearch`).
 * @param locale (string) The locale of the current user. Must comply with IETF’s BCP 47 definition: https://www.rfc-editor.org/rfc/bcp/bcp47.txt.
 */
export const updateSearchConfiguration = createAction(
  'configuration/updateSearchConfiguration',
  (payload: {
    apiBaseUrl?: string;
    pipeline?: string;
    searchHub?: string;
    locale?: string;
  }) =>
    validatePayload(payload, {
      apiBaseUrl: new StringValue({url: true, emptyAllowed: false}),
      pipeline: new StringValue({emptyAllowed: false}),
      searchHub: new StringValue({emptyAllowed: false}),
      locale: localeValidation,
    })
);

/**
 * Updates the analytics configuration.
 * @param enabled (boolean) Whether to enable usage analytics tracking.
 * @param originLevel2 (string) The origin level 2 usage analytics event metadata whose value should typically be the identifier of the tab from which the usage analytics event originates (e.g., `All`).
 * @param originLevel3 (string) The origin level 3 usage analytics event metadata whose value should typically be the URL of the page that linked to the search interface that’s making the request (e.g., `https://connect.coveo.com/s/`).
 * @param apiBaseUrl (string) The Usage Analytics API base URL to use (e.g., `https://platform.cloud.coveo.com/rest/ua`).
 * @param runtimeEnvironment (IRuntimeEnvironment) The Coveo analytics runtime to use, see https://github.com/coveo/coveo.analytics.js for more info.
 */
export const updateAnalyticsConfiguration = createAction(
  'configuration/updateAnalyticsConfiguration',
  (payload: {
    enabled?: boolean;
    originLevel2?: string;
    originLevel3?: string;
    apiBaseUrl?: string;
    runtimeEnvironment?: IRuntimeEnvironment;
  }) =>
    validatePayload(payload, {
      enabled: new BooleanValue({default: true}),
      originLevel2: originSchemaOnConfigUpdate(),
      originLevel3: originSchemaOnConfigUpdate(),
      apiBaseUrl: new StringValue({url: true, emptyAllowed: false}),
      runtimeEnvironment: new Value(),
    })
);

/**
 * Renews the accessToken specified in the global headless engine configuration.
 * @param renew (`() => Promise<string>`) A function that fetches a new access token. The function must return a Promise that resolves to a string (the new access token).
 */
export const renewAccessToken = createAsyncThunk(
  'configuration/renewAccessToken',
  async (renew: () => Promise<string>) => {
    return await renew();
  }
);

/**
 * Disables analytics tracking.
 */
export const disableAnalytics = createAction('configuration/analytics/disable');
/**
 * Enables analytics tracking.
 */
export const enableAnalytics = createAction('configuration/analytics/enable');

/**
 * Sets originLevel2 for analytics tracking.
 * @param originLevel2 (string) The origin level 2 usage analytics event metadata whose value should typically be the identifier of the tab (e.g., `All`).
 */
export const setOriginLevel2 = createAction(
  'configuration/analytics/originlevel2',
  (payload: {originLevel2: string}) =>
    validatePayload(payload, {originLevel2: originSchemaOnUpdate()})
);

/**
 * Sets originLevel3 for analytics tracking.
 * @param originLevel3 (string) The origin level 3 usage analytics event metadata whose value should typically be the URL of the page that linked to the search interface (e.g., `https://connect.coveo.com/s/`).
 */
export const setOriginLevel3 = createAction(
  'configuration/analytics/originlevel3',
  (payload: {originLevel3: string}) =>
    validatePayload(payload, {originLevel3: originSchemaOnUpdate()})
);
