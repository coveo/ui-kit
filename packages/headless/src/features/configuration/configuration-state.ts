import {platformUrl} from '../../api/platform-client';
import {IRuntimeEnvironment} from 'coveo.analytics';

export interface ConfigurationState {
  /**
   * The unique identifier of the target Coveo Cloud organization (e.g., `mycoveocloudorganizationg8tp8wu3`)
   */
  organizationId: string;
  /**
   * The access token to use to authenticate requests against the Coveo Cloud endpoints. Typically, this will be an API key or search token that grants the privileges to execute queries and push usage analytics data in the target Coveo Cloud organization.
   */
  accessToken: string;
  /**
   * The Plaform URL to use.
   * By default, https://platform.cloud.coveo.com
   */
  platformUrl: string;
  /**
   * The global headless engine Search API configuration.
   */
  search: {
    /**
     * The Search API base URL to use.
     * By default, will append /rest/search/v2 to the platformUrl value.
     */
    apiBaseUrl: string;
    /**
     * The locale of the current user. Must comply with IETF’s BCP 47 definition: https://www.rfc-editor.org/rfc/bcp/bcp47.txt.
     */
    locale: string;
  };
  /**
   * The global headless engine Usage Analytics API configuration.
   */
  analytics: {
    /**
     * Specifies if analytics tracking should be enabled. By default analytics events are tracked.
     */
    enabled: boolean;
    /**
     * The Analytics API base URL to use.
     * By default, will append /rest/ua to the platformUrl value.
     */
    apiBaseUrl: string;
    /**
     * Origin level 2 is a usage analytics event metadata whose value should typically be the name/identifier of the tab from which the usage analytics event originates.
     *
     * When logging a Search usage analytics event, originLevel2 should always be set to the same value as the corresponding tab (parameter) Search API query parameter so Coveo Machine Learning models function properly, and usage analytics reports and dashboards are coherent.
     *
     * This value is optional, and will automatically try to resolve itself from the tab search parameter.
     */
    originLevel2: string;

    /**
     * Origin level 3 is a usage analytics event metadata whose value should typically be the URL of the page that linked to the search interface that’s making the request.
     *
     * When logging a Search usage analytics event, originLevel3 should always be set to the same value as the corresponding referrer Search API query parameter so usage analytics reports and dashboards are coherent.
     *
     * This value is optional, and will automatically try to resolve itself from the referrer search parameter.
     */
    originLevel3: string;
    /**
     * Optional analytics runtime environment, this is needed for analytics to work correctly if you're running outside of a browser.
     * See https://github.com/coveo/coveo.analytics.js for more info
     */
    runtimeEnvironment?: IRuntimeEnvironment;
  };
}

export const searchAPIEndpoint = '/rest/search/v2';
export const analyticsAPIEndpoint = '/rest/ua';

export const getConfigurationInitialState: () => ConfigurationState = () => ({
  organizationId: '',
  accessToken: '',
  platformUrl: platformUrl(),
  search: {
    apiBaseUrl: `${platformUrl()}${searchAPIEndpoint}`,
    locale: 'en-US',
  },
  analytics: {
    enabled: true,
    apiBaseUrl: `${platformUrl()}${analyticsAPIEndpoint}`,
    originLevel2: 'default',
    originLevel3: 'default',
  },
});
