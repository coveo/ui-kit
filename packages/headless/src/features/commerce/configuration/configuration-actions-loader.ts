import {PayloadAction} from '@reduxjs/toolkit';
import {CommerceEngine} from '../../../app/commerce-engine/commerce-engine';
import {configurationReducer as configuration} from '../../configuration/configuration-slice';
import {
  UpdateAnalyticsConfigurationPayload,
  UpdateBasicConfigurationPayload,
  UpdateProxyBaseUrlPayload,
  disableAnalytics,
  enableAnalytics,
  updateAnalyticsConfiguration,
  updateBasicConfiguration,
  updateProxyBaseUrl,
} from './configuration-actions';

export type {
  UpdateAnalyticsConfigurationPayload,
  UpdateBasicConfigurationPayload,
};

/**
 * The configuration action creators.
 */
export interface ConfigurationActionCreators {
  /**
   * Disables analytics.
   *
   * @returns A dispatchable action.
   */
  disableAnalytics(): PayloadAction<void>;

  /**
   * Enables analytics.
   *
   * @returns A dispatchable action.
   */
  enableAnalytics(): PayloadAction<void>;

  /**
   * Updates the analytics configuration.
   *
   * @param payload - The action creator payload.
   * @returns A dispatchable action.
   */
  updateAnalyticsConfiguration(
    payload: UpdateAnalyticsConfigurationPayload
  ): PayloadAction<UpdateAnalyticsConfigurationPayload>;

  /**
   * Updates the basic configuration.
   *
   * @param payload - The action creator payload.
   * @returns A dispatchable action.
   */
  updateBasicConfiguration(
    payload: UpdateBasicConfigurationPayload
  ): PayloadAction<UpdateBasicConfigurationPayload>;

  /**
   * Updates the commerce configuration.
   *
   * @param payload - The action creator payload.
   * @returns A dispatchable action.
   */
  updateProxyBaseUrl(
    payload: UpdateProxyBaseUrlPayload
  ): PayloadAction<UpdateProxyBaseUrlPayload>;
}

/**
 * Loads the configuration reducer and returns the available commerce configuration action creators.
 *
 * @param engine - The commerce engine.
 * @returns An object holding the commerce configuration action creators.
 */
export function loadConfigurationActions(
  engine: CommerceEngine
): ConfigurationActionCreators {
  engine.addReducers({configuration});
  return {
    disableAnalytics,
    enableAnalytics,
    updateAnalyticsConfiguration,
    updateBasicConfiguration,
    updateProxyBaseUrl,
  };
}
