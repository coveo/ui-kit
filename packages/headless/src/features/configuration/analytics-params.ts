import {EventDescription} from 'coveo.analytics';
import {getAnalyticsSource} from '../../api/analytics/analytics-selectors';
import {getVisitorID} from '../../api/analytics/coveo-analytics-utils';
import {getPageID} from '../../api/analytics/search-analytics';
import {AnalyticsParam} from '../../api/search/search-api-params';
import {AnalyticsState} from './configuration-state';

export const fromAnalyticsStateToAnalyticsParams = async (
  s: AnalyticsState,
  eventDescription?: EventDescription
): Promise<AnalyticsParam> => {
  const isNextAnalytics = s.analyticsMode === 'next';
  return {
    analytics: {
      clientId: await getVisitorID(s),
      clientTimestamp: new Date().toISOString(),
      documentReferrer: s.originLevel3,
      originContext: s.originContext,
      ...(eventDescription && {
        actionCause: eventDescription.actionCause,
        customData: eventDescription.customData,
      }),
      ...(s.userDisplayName && {userDisplayName: s.userDisplayName}),
      ...(s.documentLocation && {documentLocation: s.documentLocation}),
      ...(s.deviceId && {deviceId: s.deviceId}),
      ...(getPageID() && {pageId: getPageID()}),
      ...(isNextAnalytics && s.trackingId && {trackingId: s.trackingId}),
      ...{capture: isNextAnalytics},
      ...(isNextAnalytics && {source: getAnalyticsSource(s)}),
    },
  };
};
