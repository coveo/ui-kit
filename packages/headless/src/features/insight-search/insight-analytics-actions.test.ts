import {createRelay} from '@coveo/relay';
import {InsightEngine} from '../../app/insight-engine/insight-engine';
import {ThunkExtraArguments} from '../../app/thunk-extra-arguments';
import {buildMockInsightEngine} from '../../test/mock-engine-v2';
import {buildMockInsightState} from '../../test/mock-insight-state';
import {clearMicrotaskQueue} from '../../test/unit-test-utils';
import {getConfigurationInitialState} from '../configuration/configuration-state';
import {
  logExpandToFullUI,
  logInsightCreateArticle,
} from './insight-analytics-actions';

const mockLogCreateArticle = jest.fn();
const mockLogExpandtoFullUI = jest.fn();
const emit = jest.fn();

jest.mock('@coveo/relay');

jest.mock('coveo.analytics', () => {
  const mockCoveoInsightClient = jest.fn(() => ({
    disable: jest.fn(),
    logExpandToFullUI: mockLogExpandtoFullUI,
    logCreateArticle: mockLogCreateArticle,
  }));

  return {
    CoveoInsightClient: mockCoveoInsightClient,
    history: {HistoryStore: jest.fn()},
  };
});

jest.mocked(createRelay).mockReturnValue({
  emit,
  getMeta: jest.fn(),
  on: jest.fn(),
  off: jest.fn(),
  updateConfig: jest.fn(),
  clearStorage: jest.fn(),
  version: 'foo',
});

const exampleSubject = 'example subject';
const exampleDescription = 'example description';
const exampleCaseId = '1234';
const exampleCaseNumber = '5678';
const exampleCreateArticleMetadata = {
  articleType: 'Knowledge__kav',
};

describe('insight analytics actions', () => {
  let engine: InsightEngine;
  const caseContextState = {
    caseContext: {
      Case_Subject: exampleSubject,
      Case_Description: exampleDescription,
    },
    caseId: exampleCaseId,
    caseNumber: exampleCaseNumber,
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('when analyticsMode is `legacy`', () => {
    beforeEach(() => {
      engine = buildMockInsightEngine(
        buildMockInsightState({
          configuration: {
            ...getConfigurationInitialState(),
            analytics: {
              ...getConfigurationInitialState().analytics,
              analyticsMode: 'legacy',
            },
          },
          insightCaseContext: caseContextState,
        })
      );
    });

    describe('logCreateArticle', () => {
      it('should call coveo.analytics.logCreateArticle properly', async () => {
        await logInsightCreateArticle(exampleCreateArticleMetadata)()(
          engine.dispatch,
          () => engine.state,
          {} as ThunkExtraArguments
        );

        const expectedPayload = {
          caseContext: {
            Case_Subject: exampleSubject,
            Case_Description: exampleDescription,
          },
          caseId: exampleCaseId,
          caseNumber: exampleCaseNumber,
        };

        expect(mockLogCreateArticle).toHaveBeenCalledTimes(1);
        expect(mockLogCreateArticle.mock.calls[0][0]).toStrictEqual(
          exampleCreateArticleMetadata
        );
        expect(mockLogCreateArticle.mock.calls[0][1]).toStrictEqual(
          expectedPayload
        );
      });
    });

    describe('logExpandToFullUI', () => {
      it('should call coveo.analytics.logExpandToFullUI properly', async () => {
        await logExpandToFullUI(
          exampleCaseId,
          exampleCaseNumber,
          'c__FullSearch',
          'openFullSearchButton'
        )()(engine.dispatch, () => engine.state, {} as ThunkExtraArguments);

        const expectedPayload = {
          caseContext: {
            Case_Subject: exampleSubject,
            Case_Description: exampleDescription,
          },
          caseId: exampleCaseId,
          caseNumber: exampleCaseNumber,
          fullSearchComponentName: 'c__FullSearch',
          triggeredBy: 'openFullSearchButton',
        };

        expect(mockLogExpandtoFullUI).toHaveBeenCalledTimes(1);
        expect(mockLogExpandtoFullUI.mock.calls[0][0]).toStrictEqual(
          expectedPayload
        );
      });
    });
  });

  describe('when analyticsMode is `next`', () => {
    beforeEach(() => {
      engine = buildMockInsightEngine(
        buildMockInsightState({
          configuration: {
            ...getConfigurationInitialState(),
            analytics: {
              ...getConfigurationInitialState().analytics,
              analyticsMode: 'next',
            },
          },
          insightCaseContext: caseContextState,
        })
      );
    });

    describe('logCreateArticle', () => {
      it('should call relay.emit properly', async () => {
        await logInsightCreateArticle(exampleCreateArticleMetadata)()(
          engine.dispatch,
          () => engine.state,
          {} as ThunkExtraArguments
        );
        await clearMicrotaskQueue();

        expect(emit).toHaveBeenCalledTimes(1);
        expect(emit.mock.calls[0]).toMatchSnapshot();
      });
    });

    describe('logExpandToFullUI', () => {
      it('should call relay.emit properly', async () => {
        await logExpandToFullUI(exampleCaseId, exampleCaseNumber, '', '')()(
          engine.dispatch,
          () => engine.state,
          {} as ThunkExtraArguments
        );
        await clearMicrotaskQueue();

        expect(emit).toHaveBeenCalledTimes(1);
        expect(emit.mock.calls[0]).toMatchSnapshot();
      });
    });
  });
});
