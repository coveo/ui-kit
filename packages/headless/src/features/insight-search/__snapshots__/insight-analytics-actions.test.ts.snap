// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insight analytics actions when analyticsMode is \`next\` logCreateArticle should call relay.emit properly 1`] = `
[
  "InsightPanel.CreateArticle",
  {
    "articleType": "Knowledge__kav",
    "context": {
      "caseNumber": "5678",
      "targetId": "1234",
      "targetType": "Case",
    },
  },
]
`;

exports[`insight analytics actions when analyticsMode is \`next\` logExpandToFullUI should call relay.emit properly 1`] = `
[
  "InsightPanel.ExpandToFullUI",
  {
    "context": {
      "caseNumber": "5678",
      "targetId": "1234",
      "targetType": "Case",
    },
  },
]
`;
